### ISSUE TITLE
C# DeviceClient returns multiple unexpected ConnectionStatusChanges events when the network goes down

### ISSUE THREAD

**Issue by varunpuranik** [Azure/azure-iot-sdk-csharp#211]
OS: Windows 10
SDK: v1.5.2

### Description of the issue:
If I set the ConnectionStatusChangesHandler on the DeviceClient (over AMQP), and the network goes down, I expect one call to the callback with ConnectionStatus = Disconnected/Disconnected_Retrying, and ConnectionStatusChangeReason = No_Network.
What I get instead are 3 callbacks in quick succession - 
ConnectionStatus = Disconnected_Retrying, ConnectionStatusChangeReason = No_Network
ConnectionStatus = Disabled, ConnectionStatusChangeReason = Client_Close
ConnectionStatus = Disabled, ConnectionStatusChangeReason = Client_Close

Note, that the first callback takes a long time to come through and then the next calls come quickly.
I am not sure why I am getting these additional callbacks. 

### Console log of the issue:
Note: My app is a Dotnet Core 2.0 app on VS2017

```
private static async Task<DeviceClient> TestCallback()
{
            string connectionString = "My connection string"
            var dc = DeviceClient.CreateFromConnectionString(connectionString);            
            await dc.OpenAsync();
            dc.SetConnectionStatusChangesHandler(ConnectionStatusChanges);
            await dc.SetDesiredPropertyUpdateCallbackAsync(DesiredPropertiesUpdates, null);            
            
            return dc;
}

private static void ConnectionStatusChanges(ConnectionStatus status, ConnectionStatusChangeReason reason)
{
            Console.WriteLine($"Connection status changed - ConnectionStatus = {status}, ConnectionStatusChangeReason = {reason}");
}

private static Task DesiredPropertiesUpdates(TwinCollection desiredProperties, object userContext)
{
            Console.WriteLine("Desired properties updated!");
            return Task.CompletedTask;
}
```

###Console Log:
Connection status changed - ConnectionStatus = Disconnected_Retrying, ConnectionStatusChangeReason = No_Network
Connection status changed - ConnectionStatus = Disabled, ConnectionStatusChangeReason = Client_Close
Connection status changed - ConnectionStatus = Disabled, ConnectionStatusChangeReason = Client_Close

**Comment by ryan-z-johnson** [comment]
Thanks for reporting, I am able to reproduce your issue. I'm looking into a fix now.

**Comment by rquackenbush** [comment]
I'm seeing the same thing in my environment. I'm once this happens, the DeviceClient will never reconnect without restarting the host process.

Any updates?

My environment:
- .NET Core SDK 2.1.300
- `Microsoft.Azure.Devices.Client` Version="1.17.0"
- Windows 10 / Linux (Raspberry Pi running resin-os / in a container)

**Comment by laorient** [comment]
Me too. I am on .NET Standard 2.0. Microsoft.Azure.Devices.Client 1.17.1, Windows 10/PC. It happens when using Amqp.

**Comment by CIPop** [comment]
We are investigating multiple issues around Retries in DeviceClient/ModuleClient.

At the moment we believe that the feature is not correctly implemented and requires some portions to be investigated and potentially rewritten. We have 3 developers working on similar issues at the moment.

There is no simple workaround – given that the only way to retrieve the information is in an re-entrant callback: I would try to dispose of the client first then reconnect within the application code.

**Comment by pangar-ms** [comment]
Any update on this issue, this bug is active since long.

**Comment by prmathur-microsoft** [comment]
Yes. We have uncovered major design changes related to issues like this. We are working on a  redesign to fix these issues. We will keep you updated. However at the moment, expect to have fixes by End of January.

**Comment by sebader** [comment]
I ran into a similar situation within a custom module for IoT Edge (ModuleClient):
```
2019-01-08 08:49:01.511 +00:00 [INF] - Module connection changed. New status=Disconnected_Retrying Reason=Communication_Error
2019-01-08 08:53:01.544 +00:00 [INF] - Module connection changed. New status=Disconnected Reason=Retry_Expired
```

By the time the second message came, the Edge Hub was in fact up and running again. Is this behaviour part of this bug here or what is the proper way to handle this? 
Should you exit the module once you get a "Retry_Expired" so that the EdgeAgent will restart it again (and again...)?

**Comment by trbenning** [comment]
It looks like the retry handling underwent a significant refactoring in #741, which was released in v1.19.0. Do those changes address this issue, or is more work planned to address this?

**Comment by prmathur-microsoft** [comment]
More work is on the way with significant design changes to AMQP. However this initial set of changes should address any retry failures you were seeing. Could you give it a try and let us know if you still see issues?

**Comment by trbenning** [comment]
Yes, of course. We'll update the SDK to 1.19.0 on a few of our test devices, and I'll report back in a few days on what we see.

**Comment by prmathur-microsoft** [comment]
This is currently work in progress and release is trending to first week of April. I'll update here with the release date as soon as I have it.

**Comment by CIPop** [comment]
@sebader I believe you are hitting #975 which is in the common transport code (MQTT/AMQP).

Duplicate events have been solved in 1.19.0

**Comment by az-iot-builder-01** [comment]
@rquackenbush, @laorient, @CIPop, @pangar-ms, @prmathur-microsoft, @sebader, @trbenning, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://forms.office.com/Pages/ResponsePage.aspx?id=v4j5cvGGr0GRqy180BHbRx8jpG79OudPqY-BojcKjkJUMFFBQjgxNkw5ODJIRjBWTzI0Nk85R0pUQiQlQCN0PWcu)
